File: 14. Longest Common Prefix.go
package main

import "fmt"

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func countSmallestChar(strs []string) int {
	smallestLen := len(strs[0])
	for i := 0; i < len(strs); i++ {
		smallestLen = min(smallestLen, len(strs[i]))
	}
	return smallestLen
}

func longestCommonPrefix(strs []string) string {
	answer := ""

	minLen := countSmallestChar(strs)

	for i := 0; i < minLen; i++ {
		currentChar := strs[0][i]
		for j := 1; j < len(strs); j++ {
			if strs[j][i] != currentChar {
				return answer
			}
		}
		answer = answer + string(currentChar)
	}

	return answer
}

func main() {

	strs := []string{"flower", "flow", "flight"}
	answer := longestCommonPrefix(strs)
	fmt.Println("Answer: ", answer)
}

File: 14. Longest Common Prefix.py
from typing import List

class Node:
    
    def __init__(self):
        self.children = {}
        self.isEnd = False

class TrieSolution:

    def __init__(self):
        self.result = ""
        self.root = Node()

    def insertIntoTrie(self,word: str):
        cur = self.root
        for ch in word:
            if ch in cur.children: 
                cur = cur.children[ch]
                continue
            else: cur.children[ch] = Node()
            cur = cur.children[ch]
        cur.isEnd = True

    def longestCommonPrefix(self, strs: List[str]) -> str:

        for word in strs:
            if word == "": return ""
            self.insertIntoTrie(word)
        
        # Traverse trie root and find answer
        cur = self.root
        while len(cur.children) == 1:
            val = list(cur.children)[0]
            if cur.isEnd: break
            self.result += val
            cur = cur.children[val]
        
        return self.result


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        
        result = ""
        cur = ""
        minStr = 200
        for s in strs:
            minStr = min(minStr,len(s))

        for i in range(minStr):
            cur = strs[0][i]
            for word in strs:
                if word[i] != cur:
                    return result
            
            result = result + cur
        return result